README.md
# Simple SLS SSR

A simple helper package to turn Serverless Framework into a Server Side Rendering Application. Uses a very basic Templating Engine for building application pages.

## Quickstart use within a Lambda

### Setup
- Create a folder in your root named ``/templates``.
- Create your html files to use as templates in the ``/templates`` folder.
- Create a folder in your root named ``/styles``.
- Create your css files in the ``/styles`` folder.
- Name your css files the same as your html files (file extension will be different of course)

### Serving a page
```javascript
const { render } = require('simple-sls-ssr')

// Your Lambda
module.exports.myfunction = async (event, context) => {
  return await render('dashboard')
}
```

This will render a page in the templates folder named ``dashboard.html``.

### Serving a page with variables
Use handlebars to declare your variable name.

hello.html
```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Hello World</title>
</head>
<body>
  <h1>Hello {{name}}</h1>
</body>
</html>
```

```javascript
module.exports.myfunction = async (event, context) => {
  const variables = {
    name: "Jane"
  }

  return await render('hello', variables)
}
```

### Setting Custom Headers
```javascript
module.exports.myfunction = async (event, context) => {
  const headers = {
    'Content-Type': 'application/json'
  }

  return await render('hello', null, headers)
}
```

### Setting Custom Style Sheet
The render function will look for the css file in the `styles` folder that is named the same as the view.
```javascript
module.exports.myfunction = async (event, context) => {
  const headers = {
    'Content-Type': 'application/json'
  }

  return await render('hello', null, headers, true)
}
```

### Redirecting to another page
```javascript
const { response } = require('simple-sls-ssr')

module.exports.myfunction = async (event, context) => {
  const relativePath = './login'
  return response.redirect(relativePath)
}
```

### Returning non templated, http responses
```javascript
const { response } = require('simple-sls-ssr')

module.exports.myfunction = async (event, context) => {
  const body = { success: "hello world!" }
  return response.http(200, body)
}
```


CHANGELOG.md
# CHANGELOG

## v1.1.0
- Adds ability to include styles from css file

## v.1.0.2
- Adds Stricter security headers

tests/render.tests.js
'use strict'

const test = require('tape')

const render = require('../modules/render')

test('Test render of basic page', async t => {
  const response = await render('render')

  const expected = {
    headers: {
      'Content-Type': 'text/html',
      'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload',
      'Content-Security-Policy': 'default-src "self"',
      'X-Frame-Options': 'deny',
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': 'origin-when-cross-origin',
      'Cache-Control': 'no-store',
      'Clear-Site-Data': '*'
    },
    statusCode: 200,
    body: '<!DOCTYPE html>\n<html lang="en">\n<head>\n  <meta charset="UTF-8">\n  <meta http-equiv="X-UA-Compatible" content="IE=edge">\n  <meta name="viewport" content="width=device-width, initial-scale=1.0">\n  <title>Document</title>\n</head>\n<body>\n  \n</body>\n</html>'
  }

  t.deepEquals(response, expected, 'Renders the correct information.')

  t.end()
})

test('Test render of a page with variables', async t => {
  const variables = { name: 'Jane' }
  const response = await render('render-variables', variables)

  const expected = {
    headers: {
      'Content-Type': 'text/html',
      'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload',
      'Content-Security-Policy': 'default-src "self"',
      'X-Frame-Options': 'deny',
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': 'origin-when-cross-origin',
      'Cache-Control': 'no-store',
      'Clear-Site-Data': '*'
    },
    statusCode: 200,
    body: '<!DOCTYPE html>\n<html lang="en">\n<head>\n  <meta charset="UTF-8">\n  <meta http-equiv="X-UA-Compatible" content="IE=edge">\n  <meta name="viewport" content="width=device-width, initial-scale=1.0">\n  <title>Document</title>\n</head>\n<body>\n  <h1>Hi Jane</h1>\n</body>\n</html>'
  }

  t.deepEquals(response, expected, 'Renders the correct information with the variables.')

  t.end()
})

test('Test render of a page with variables', async t => {
  const headers = { 'Set-Cookie': 'mycookie=123' }
  const response = await render('render-headers', null, headers)

  const expected = {
    headers: {
      'Content-Type': 'text/html',
      'Set-Cookie': 'mycookie=123',
      'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload',
      'Content-Security-Policy': 'default-src "self"',
      'X-Frame-Options': 'deny',
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': 'origin-when-cross-origin',
      'Cache-Control': 'no-store',
      'Clear-Site-Data': '*'
    },
    statusCode: 200,
    body: '<!DOCTYPE html>\n<html lang="en">\n<head>\n  <meta charset="UTF-8">\n  <meta http-equiv="X-UA-Compatible" content="IE=edge">\n  <meta name="viewport" content="width=device-width, initial-scale=1.0">\n  <title>Document</title>\n</head>\n<body>\n  \n</body>\n</html>'
  }

  t.deepEquals(response, expected, 'Renders the correct information with the variables.')

  t.end()
})

test('Errors when template requires a variable it cannot find', async t => {
  try {
    await render('render-variables')

    t.fail('Should have complained that there is no variable')
  } catch (error) {
    t.equals(error.toString(), 'ReferenceError: name is not defined', 'We get the correct error')
    t.end()
  }
})

test('Should render with css', async t => {
  const headers = { 'Set-Cookie': 'mycookie=123' }
  const response = await render('render-css', null, headers, true)

  const expected = {
    headers: {
      'Content-Type': 'text/html',
      'Set-Cookie': 'mycookie=123',
      'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload',
      'Content-Security-Policy': 'default-src "self"',
      'X-Frame-Options': 'deny',
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': 'origin-when-cross-origin',
      'Cache-Control': 'no-store',
      'Clear-Site-Data': '*'
    },
    statusCode: 200,
    body: '<!DOCTYPE html>\n<html lang="en">\n<head>\n  <meta charset="UTF-8">\n  <meta http-equiv="X-UA-Compatible" content="IE=edge">\n  <meta name="viewport" content="width=device-width, initial-scale=1.0">\n  <style>\n    body {\n  color: gray;\n}\n  </style>\n  <title>Document</title>\n</head>\n<body>\n</body>\n</html>'
  }

  t.deepEquals(response, expected, 'Renders the correct information with the css.')

  t.end()
})

test('Should not render css when style file is not found', async t => {
  const headers = { 'Set-Cookie': 'mycookie=123' }
  const response = await render('render-css-no-style-file', null, headers, true)

  const expected = {
    headers: {
      'Content-Type': 'text/html',
      'Set-Cookie': 'mycookie=123',
      'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload',
      'Content-Security-Policy': 'default-src "self"',
      'X-Frame-Options': 'deny',
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': 'origin-when-cross-origin',
      'Cache-Control': 'no-store',
      'Clear-Site-Data': '*'
    },
    statusCode: 200,
    body: '<!DOCTYPE html>\n<html lang="en">\n<head>\n  <meta charset="UTF-8">\n  <meta http-equiv="X-UA-Compatible" content="IE=edge">\n  <meta name="viewport" content="width=device-width, initial-scale=1.0">\n  <style>\n    \n  </style>\n  <title>Document</title>\n</head>\n<body>\n</body>\n</html>'
  }

  t.deepEquals(response, expected, 'Renders the correct information with the css.')

  t.end()
})


tests/response.tests.js
'use strict'

const test = require('tape')

const { http, redirect } = require('../modules/response')

test('Test generate return data with 200 htttp code.', t => {
  const response = http(200)

  const expected = {
    statusCode: 200,
    headers: {
      'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload',
      'Content-Security-Policy': 'default-src "self"',
      'X-Frame-Options': 'deny',
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': 'origin-when-cross-origin',
      'Cache-Control': 'no-store',
      'Clear-Site-Data': '*'
    }
  }

  t.deepEquals(response, expected, 'Renders the correct information.')

  t.end()
})

test('Test generate return data with 200 htttp code and body', t => {
  const body = {
    success: 'Hello'
  }
  const response = http(200, body)

  const expected = {
    statusCode: 200,
    headers: {
      'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload',
      'Content-Security-Policy': 'default-src "self"',
      'X-Frame-Options': 'deny',
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': 'origin-when-cross-origin',
      'Cache-Control': 'no-store',
      'Clear-Site-Data': '*'
    },
    body: '{"success":"Hello"}'
  }

  t.deepEquals(response, expected, 'Renders the correct information.')

  t.end()
})

test('Test generating a redirect response.', async t => {
  const response = await redirect('./mypage')

  const expected = {
    statusCode: 301,
    headers: {
      Location: './mypage',
      'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload',
      'Content-Security-Policy': 'default-src "self"',
      'X-Frame-Options': 'deny',
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': 'origin-when-cross-origin',
      'Cache-Control': 'no-store',
      'Clear-Site-Data': '*'
    }
  }

  t.deepEquals(response, expected, 'Renders the correct information.')

  t.end()
})

test('Test generating a redirect response with custom headers.', async t => {
  const headers = {
    'Set-Cookie': 'mycookie=123'
  }
  const response = await redirect('./mypage', headers)

  const expected = {
    statusCode: 301,
    headers: {
      Location: './mypage',
      'Set-Cookie': 'mycookie=123',
      'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload',
      'Content-Security-Policy': 'default-src "self"',
      'X-Frame-Options': 'deny',
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': 'origin-when-cross-origin',
      'Cache-Control': 'no-store',
      'Clear-Site-Data': '*'
    }
  }

  t.deepEquals(response, expected, 'Renders the correct information.')

  t.end()
})


modules/response.js
'use strict'

/**
 * @description Formats the http response for the http functions.
 * @param {integer} statusCode
 * @param {object} body
 * @returns {object} Http return data.
 */
module.exports.http = (statusCode, body) => {
  const returnData = {
    statusCode: statusCode,
    headers: {
      'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload',
      'Content-Security-Policy': 'default-src "self"',
      'X-Frame-Options': 'deny',
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': 'origin-when-cross-origin',
      'Cache-Control': 'no-store',
      'Clear-Site-Data': '*'
    }
  }

  if (body) {
    returnData.body = JSON.stringify(body)
  }
  return returnData
}

/**
 * @description Builds response w/ a redirect to a url.
 * @param {string} redirectPath relative path to redirect to.
 * @param {object} headers Headers to add to the return object.
 * @returns {object} Http return data.
 */
module.exports.redirect = async (redirectPath, headers) => {
  const returnData = {
    statusCode: 301,
    headers: {
      Location: redirectPath,
      'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload',
      'Content-Security-Policy': 'default-src "self"',
      'X-Frame-Options': 'deny',
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': 'origin-when-cross-origin',
      'Cache-Control': 'no-store',
      'Clear-Site-Data': '*'
    }
  }

  if (headers) {
    returnData.headers = { ...returnData.headers, ...headers }
  }

  return returnData
}


modules/render.js
'use strict'

const _ = require('lodash')
const fs = require('fs')
const path = require('path')
const validator = require('validator')
const response = require('./response')

// Use double handlebars for template variables
_.templateSettings.interpolate = /{{([\s\S]+?)}}/g

/**
 * @description Simple templating helper to help us
 * display html. Use {{ key }} to bring in a template.
 * see https://docs-lodash.com/v4/template/
 * @param {string} view The html template name in the template folder.
 * @param {object} variables Variables that will be rendered into the html.
 * @param {object} headers Custom html headers you want to add.
 * @returns {object} Http return information with
 * template rendered for display.
 */
module.exports = async (view, variables = {}, headers = {}, hasStylesheet = false) => {
  const returnData = {
    headers: {
      'Content-Type': 'text/html',
      'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload',
      'Content-Security-Policy': 'default-src "self"',
      'X-Frame-Options': 'deny',
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': 'origin-when-cross-origin',
      'Cache-Control': 'no-store',
      'Clear-Site-Data': '*'
    },
    statusCode: 200
  }

  if (Object.keys(headers).length) {
    returnData.headers = { ...returnData.headers, ...headers }
  }

  const templateVariables = {}
  if (variables) {
    for (const [key, value] of Object.entries(variables)) {
      templateVariables[key] = validator.escape(value.toString())
    }
  }

  if (hasStylesheet) {
    try {
      const file = await fs.readFileSync(path.resolve(
        'styles',
        `${view}.css`
      ), 'utf-8')
      templateVariables.styles = file
    } catch (error) {
      templateVariables.styles = ''
    }
  }

  try {
    returnData.body = await _.template(
      fs.readFileSync(
        path.resolve(
          'templates',
          `${view}.html`
        )
      )
    )(templateVariables)
  } catch (error) {
    returnData.body = await _.template(
      fs.readFileSync(
        path.resolve(
          'templates',
          `${view}.html`
        )
      )
    )(templateVariables)
    return response.error(400)
  }

  return returnData
}
